ReSharper.FSharp/src/FSharp/FSharp.Psi.Intentions/src/Intentions/FunctionAnnotationAction.fs

```diff
@@ -35,6 +35,60 @@ module SpecifyTypes =
         static member ParameterTypesOnly = { CanSpecifyParameterTypes = true; CanSpecifyReturnType = false }
         static member ReturnTypeOnly = { CanSpecifyParameterTypes = false; CanSpecifyReturnType = true }
 
+    let rec private collectTypeUsages acc (context, fcsType: FSharpType, typeUsage: ITypeUsage) =
+        if fcsType.IsGenericParameter then acc else
+
+        match typeUsage with
+        | :? INamedTypeUsage as typeUsage ->
+            let typeReference = typeUsage.ReferenceName
+            let acc = (typeUsage, fcsType, context) :: acc
+
+            let typeArgs = typeReference.TypeArgumentList
+            if isNull typeArgs then acc else
+
+            typeArgs.TypeUsagesEnumerable
+            |> Seq.mapi (fun i typeArg -> i, typeArg)
+            |> Seq.fold (fun acc (i, typeArg) ->
+                let fcsType = fcsType.GenericArguments[i]
+                collectTypeUsages acc (context, fcsType, typeArg)) acc
+
+        | :? ITupleTypeUsage as typeUsage ->
+            typeUsage.Items
+            |> Seq.mapi (fun i field -> i, field)
+            |> Seq.fold (fun acc (i, typeUsage) -> collectTypeUsages acc (context, fcsType.GenericArguments[i], typeUsage)) acc
+
+        | :? IParenTypeUsage as typeUsage ->
+            collectTypeUsages acc (context, fcsType, typeUsage.InnerTypeUsage)
+
+        | :? IFunctionTypeUsage as typeUsage ->
+            let fcsType =
+                if fcsType.IsFunctionType then fcsType else getAbbreviatedType fcsType
+
+            let argTypeUsage = typeUsage.ArgumentTypeUsage
+            let argType = fcsType.GenericArguments[0]
+            let acc = collectTypeUsages acc (context, argType, argTypeUsage)
+
+            let returnTypeUsage = typeUsage.ReturnTypeUsage
+            let returnType = fcsType.GenericArguments[1]
+            collectTypeUsages acc (context, returnType, returnTypeUsage)
+
+        | :? IArrayTypeUsage as typeUsage ->
+            let typeUsage = typeUsage.TypeUsage
+            let fcsType = fcsType.GenericArguments[0]
+            collectTypeUsages acc (context, fcsType, typeUsage)
+
+        | :? IAnonRecordTypeUsage as typeUsage ->
+            typeUsage.Fields
+            |> Seq.mapi (fun i field -> i, field)
+            |> Seq.fold (fun acc (i, field) ->
+                let typeUsage = field.TypeUsage
+                collectTypeUsages acc (context, fcsType.GenericArguments[i], typeUsage)) acc
```

Comment:
It looks like this pattern is repeated at least three times. Would extracting a function and doing a simple `Seq.zip` over type usages and FCS type arguments work here?